jina/types/arrays/memmap.py

```diff
@@ -214,15 +326,27 @@ def __iter__(self) -> Iterator['Document']:
     def __setitem__(self, key: Union[int, str], value: 'Document') -> None:
         if isinstance(key, int):
             if 0 <= key < len(self):
+                str_key = self._int2str_id(key)
                 # override an existing entry
-                self.append(value)
-                self._header_map[self._int2str_id(key)] = self._header_map[value.id]
-                del self[value.id]
+                self.update(value, key)
+
+                # allows overwriting an existing document
+                if str_key != value.id:
+                    entry = self._header_map.pop(value.id)
+                    self._header_map = OrderedDict(
+                        [
+                            (value.id, entry) if k == str_key else (k, v)
+                            for k, v in self._header_map.items()
+                        ]
+                    )
+                    if str_key in self.buffer_pool.doc_map:
+                        self.buffer_pool.doc_map.pop(str_key)
             else:
                 raise IndexError(f'`key`={key} is out of range')
         elif isinstance(key, str):
-            value.id = key
-            self.append(value)
+            if key != value.id:
+                raise ValueError('key must be equal to document id')
+            self.update(value, key)
```

Comment:
Update expects an `idx` and not a `key`.

Reply:
fixed